Add Hugo Deployment Guide (#504)
zeit/docs: pages/guides/deploying-hugo-with-now.mdx

import Guide from '~/components/layout/guide'
import { TerminalInput } from '~/components/text/terminal'
import Caption from '~/components/text/caption'
import { InlineCode } from '~/components/text/code'
import Card from '~/components/card'
import Note from '~/components/text/note'
import { GenericLink } from '~/components/text/link'

export const meta = {
  title: 'Create a Hugo Website and Deploy It with Now',
  description:
    'Learn how to create a Hugo website and deploy it live with ZEIT Now',
  published: '2019-02-20T03:00:00.860Z',
  authors: ['unicodeveloper'],
  url: '/guides/deploying-hugo-with-now'
}

> [Hugo](https://gohugo.io) is a very popular framework for creating static websites. It's fast and flexible.

Reviewer: ```suggestion
[Hugo](https://gohugo.io) is a popular framework for generating static websites with speed and flexibility.
```

In this guide, we will set up a Hugo website and deploy it with ZEIT Now.

> ## Step 1: Set Up Your Hugo Project

Reviewer: ```suggestion
## Step 1: Setting Up Your Hugo Project
```

> To get a Hugo project running, you need to install a precompiled binary on your machine. Hugo currently provides pre-built binaries for MacOS, Windows, and Linux.

Reviewer: Let's add links where we can. Also, it's `macOS` 😄 
```suggestion
To [get a Hugo project running](https://gohugo.io/getting-started/quick-start/), you need to install a precompiled binary on your machine. Hugo currently provides [pre-built binaries for macOS, Windows, and Linux](https://gohugo.io/getting-started/installing/).
```

> Therefore, you can either download the appropriate version for your machine from the [GitHub releases](https://github.com/gohugoio/hugo/releases) or install it via your machine's package manager.

Reviewer: Perhaps mention that you can install Hugo directly last. Users want an easy method.

> For MacOS users, you can install Hugo via Brew:

Reviewer: ```suggestion
For macOS users, you can install Hugo via [Homebrew](https://brew.sh/):
```

I know you linked this below, but this is the first reference.

<TerminalInput>brew install hugo</TerminalInput>

> <Caption>Installing <GenericLink href="https://gohugo.io/getting-started/installing/">Hugo CLI</GenericLink> via <GenericLink href="https://brew.sh/">Homebrew</GenericLink>.</Caption>

Reviewer: Let's offer more than just the macOS installing method? It seems to limit the reader. Also, let's link again the installation docs to emphasize the multiple install methods.

Reviewer: Mac OS

$ brew install hugo
Windows

$ choco install hugo -confirm
Linux

$ snap install hugo

Author: Do we need to actually list all these methods out. I already made a clear sentence about how to download the appropriate versions for their different machines or use their different package managers @timothyis

Next, run the following command to create a Hugo project via the CLI:

<TerminalInput>hugo new site my-hugo-project</TerminalInput>
<Caption>Creating a new Hugo project via the CLI.</Caption>

You can add a theme to beautify the newly created site. `cd` into the `my-hugo-project` directory and run the following command to add a [theme](https://themes.gohugo.io/):

<TerminalInput>git init && git submodule add https://github.com/budparr/gohugo-theme-ananke.git themes/ananke</TerminalInput>
<Caption>Download a new theme to the Hugo project.</Caption>

The next step is to activate the theme. From your terminal, add the [ananke](https://themes.gohugo.io/gohugo-theme-ananke/) theme to the `config.toml` file.

<TerminalInput>echo 'theme = "ananke"' >> config.toml</TerminalInput>
<Caption>Edit your config.toml configuration file, and add the theme</Caption>

> ### Create Some Content

Reviewer: ```suggestion
### Creating Content for Your Project
```

You can add some content to the Hugo project by running the following command:

<TerminalInput>hugo new posts/my-first-post.md</TerminalInput>
<Caption>Creating a new post in the Hugo project via the CLI.</Caption>

> Once you are done running the command, edit the `my-first-post.md` file and add the following content:

Reviewer: ```suggestion
Once you are done running the command, edit the `my-first-post.md` file and add the following content below the generated metadata:
```
Something I paused to think about.

> ```

Reviewer: ```suggestion
```md
```

it's either md or markdown

Reviewer: Doesn't look like this was applied @unicodeveloper

Author: I didn't see this 🤦‍♂️

## Deploying Hugo with ZEIT Now
```

You can now see your project running locally with the following command:

<TerminalInput>hugo server -D</TerminalInput>
<Caption>Running the server locally with drafts enabled</Caption>

## Step 2: Deploying Your Hugo Website with Now

> Your Hugo website is ready for deployment. Create a [`now.json` configuration file](https://zeit.co/docs/v2/deployments/configuration) in the project directory and add the following content.

Reviewer: ```suggestion
With setup complete and content added, your Hugo website is ready to be deployed live.

Create a [`now.json` file](https://zeit.co/docs/v2/deployments/configuration) in the project directory and add the following content:
```

```json
{
  "version": 2,
  "name": "my-hugo-project",
  "builds": [
    {
      "src": "build.sh",
      "use": "@now/static-build",
      "config": { "distDir": "public" }
    }
  ]
}
```

<Caption>
  A <InlineCode>now.json</InlineCode> configuration file that builds a Hugo
  project.
</Caption>

> The `now.json` file allows your deployment achieve a few things:

Reviewer: ```suggestion
The `now.json` file allows you to achieve a few things with your deployment:
```

- Defines a `version` property to ensure you are using the latest [Now 2.0 platform](/docs/v2/platform/overview) version.
- Defines a project `name` that your deployments will be sorted under and known by under Now.

> - The [`builds` property](/docs/v2/deployments/builds) allows Now to use a [Builder](/docs/v2/deployments/builders/overview/) with a specific source target. We used the [`@now/static-build`](/docs/v2/deployments/official-builders/static-build-now-static-build/) to build and deploy a static project. This module takes the `build.sh` file as the entry point in this case, executes the commands inside the shell file, and exposes the resulting `public` directory as the [build output](/docs/v2/deployments/builds/#sources-and-outputs) for serving.

Reviewer: ```suggestion
- The [`builds` property](/docs/v2/deployments/builds) allows Now to use a [Builder](/docs/v2/deployments/builders/overview/) with a specific source target. We used the [`@now/static-build`](/docs/v2/deployments/official-builders/static-build-now-static-build/) to build and deploy a static project. This module takes the `build.sh` file as the entry point, executes the commands inside the shell file, and exposes the resulting `public` directory as the [build output](/docs/v2/deployments/builds/#sources-and-outputs) for serving.
```

- The [`distDir` option](/docs/v2/deployments/official-builders/static-build-now-static-build/#configuring-the-dist-directory) in the build step instructs Now to identify `public` as the static folder and build output directory.

> Now, create the `build.sh` file in the `my-hugo-project` root directory and add the following code to it:

Reviewer: ```suggestion
Now, create the `build.sh` file in the root of the `my-hugo-project` directory and add the following code to it:
```

```bash
yum install -y wget

wget https://github.com/gohugoio/hugo/releases/download/v0.31.1/hugo_0.31.1_Linux-64bit.tar.gz
tar -xzf hugo_0.31.1_Linux-64bit.tar.gz

./hugo
```

<Caption>
  A <InlineCode>build.sh</InlineCode> shell file containing commands to install
  and run Hugo on Now
</Caption>

Ensure that you run the following command on the `build.sh` file to make it executable.

> <TerminalInput>chmod +x build.sh</TerminalInput>

Reviewer: I'm not so sure that this is necessary now that we're not touching files specifically. I'll test!

Author: It's necessary

<Caption>
  A command to grant appropriate permissions to make <InlineCode>build.sh</InlineCode> executable
</Caption>

And with this configuration set up, you can now deploy your Hugo project.

<Note>

>   Finally, before deploying, ensure the <InlineCode>draft</InlineCode> value in

Reviewer: ```suggestion
  Before deploying, ensure the <InlineCode>draft</InlineCode> value in
```

  your <InlineCode>my-first-post.md</InlineCode> file is set to{' '}
  <InlineCode>false</InlineCode>. If it remains as <InlineCode>true</InlineCode>,
  the content of your posts won't show up in production.
</Note>

> If you have Now CLI on your system, you can deploy from your terminal with just one command:

Reviewer: Let's add emphasis on how to install Now CLI. Many readers won't be coming from the perspective that they know what now is.

Author: I'd just link to the NOW CLI installation link

Reviewer: I disagree. Now Desktop automatically installs Now CLI for the user and keeps it up-to-date. It's by far the best way to install Now and you still use Now CLI in the example. I realise we've made this mistake previous, now.

<TerminalInput>now</TerminalInput>

Otherwise, if you want your application to deploy automatically, you can [install the Now for GitHub app](/docs/v2/integrations/now-for-github/) and have updates for your GitHub repository deploy and alias on every push.

When your app is deployed, you will receive a deployment URL like the following: [https://my-hugo-project-kj4fu0pcq.now.sh/](https://my-hugo-project-kj4fu0pcq.now.sh/)

## Bonus: Cache Your Hugo Assets

> Caching makes your site fast and reduces bandwith consumption for your users. You can configure Now's routes within your `now.json` configuration file to set up caching headers.

Reviewer: ```suggestion
Caching makes your site fast and reduces bandwidth consumption for your users and cost for you. You can configure Now's [routes](https://zeit.co/docs/v2/deployments/routes) property within your `now.json` configuration file to set up caching headers.
```

```json
{
  ...
  "routes": [
    {"src": "^/dist/js/(.*)", "headers": {"cache-control": "public,max-age=31536000,immutable"}, "dest": "/js/$1" },
    {"src": "^/dist/css/(.*)", "headers": {"cache-control": "public,max-age=31536000,immutable"}, "dest": "/css/$1"},
    {"src": "^/images/(.*)", "headers": {"cache-control": "max-age=31536000,immutable"}, "dest": "/images/$1"},
    {"src": "^/(.*).html", "headers": {"cache-control": "public,max-age=0,must-revalidate"}, "dest": "/index.html"}
  ]
}
```

<Caption>
  Routes configuration for setting caching headers in a{' '}
  <InlineCode>now.json</InlineCode> configuration file.
</Caption>

## Resources

For more information on working with Hugo, please refer to [their documentation](https://gohugo.io/documentation/).

To configure Now further, please see these additional topics and guides:

<Card title="Deploying Basics" href="/docs/v2/deployments/basics">
  Deploy any of your applications with ZEIT Now.
</Card>

<Card
  title="Aliasing"
  href="/docs/v2/domains-and-aliases/aliasing-a-deployment/"
>
  Learn more about aliasing to your deployments.
</Card>

<Card

>   title="Deploying Gatsby with Now"

Reviewer: They came here for Hugo, let's link https://zeit.co/docs/v2/deployments/routes instead.
